### ISSUE TITLE
Parallelization of fallback index resultset retrieving

### ISSUE THREAD

**Issue by othmanemoumni88** [npgall/cqengine#255]
The fallback index use an iterator of the ObjectSet using a FilteringIterator to retrieve the resultSet of a query.
It would be great if we can choose wheather we want to use a parallelization here (using java Stream and the predicate Query#matches for example).

**Comment by npgall** [comment]
This wouldn't be so straightforward. 

It's fairly straightforward to parallelize streams which read directly from data structures which support range lookup (e.g. by index), especially when those data structures are immutable. For example, when we have an array or a List as the source, we can read the range of objects from index i and j in one thread, and read the range from index x to y in another thread - thus parallelization works well there.

**Sets are often backed by trees**
However, when dealing with Sets (which an ObjectSet is often backed by), that is not always supported. A lot of Sets are actually backed by tree or trie data structures. In those cases, when data is requested, there is typically a stateful Iterator object created, which holds a pointer to the current node within the tree or trie that is being traversed. When the application requests the next object from the iterator, the iterator will traverse to the next node(s) in the tree/trie to locate the next object. This kind of thing is very difficult to parallelize, because all threads typically would need to start traversal at the same place (e.g. the root node).

**Filtering headaches**
The situation becomes even more complicated when filtering is involved. Even if we could parallelize access to the underlying data structure (say by assigning a different segment or range within that data structure to be processed by each thread), some threads could run out of objects to process before others, if for whatever reason the number of objects within their segment which match the filter criteria, was less than the number matched in other segments. So there could be an imbalance in the number of records processed by each thread/CPU core.

**Sorting/ordering**
Finally, if the application requested results to be ordered by some arbitrary attribute (i.e. not by the index), then it would be necessary to do some kind of merge sort of the data gathered by the multiple threads at the end, to rearrange the filtered objects in ascending or descending order of the attribute. Multi-threaded sort algorithms exist, but there is usually some contention between threads. So the cost-benefit of that approach is far from clear.


**Stateful iterators**
CQEngine makes very heavy use of stateful iterators (/ResultSets) almost everywhere internally. For accessing the underlying collection (via ObjectSet), for accessing indexes, and for sorting results. It usually organizes iterators/ResultSets into a tree, where the leaf iterators in the tree can each read from different data structures (like indexes) and the parent nodes merge and sort the results. This kind of thing is very difficult to parallelize.

So even if one of the underlying data structures (say the ObjectSet, or one of the indexes) supported parallel access (and it's rare that they do), it would be very difficult to expose that functionality to the application.

**Recommended way to parallelize reads from CQEngine**
Instead, I usually recommend one of the following approaches to enable applications to process results in parallel:
1. Put your partitioning criteria into your query, and then execute multiple queries in parallel in different threads. For example if your dataset can be partitioned by range, put different non-overlapping ranges into your different queries.
2. Use a single thread to read from CQEngine, and have that thread assign the objects it reads to a pool of other threads to be processed.

Basically, CQEngine can execute multiple queries in parallel, but it doesn't support parallelism within individual queries.

Hope that helps,
Niall
